### ISSUE TITLE
SNMP Value Retrieval

### ISSUE THREAD

**Issue by wglenn01** [louislam/uptime-kuma#1675]
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

New Monitor

### 🔖 Feature description

I would love to be able to put in custom SNMP OID's per device to monitor the value of that OID with each update and display them in the table, would be SO helpful.

### ✔️ Solution

Add SNMP monitoring for custom OID's per device.

**Comment by Computroniks** [comment]
I am not sure if this is just slightly out of scope for Uptime Kuma. It feels like it would take the project closer to the likes of zabbix or checkmk as an entire system monitoring service as apposed to a system to monitor what the average user sees. However, it is @louislam descision. I would be interested to see what others thoughts are on this.

**Comment by louislam** [comment]
As long as it is about uptime and downtime, it could be a monitor type in Uptime Kuma.

But honestly, I probably won't implement this, as I don't use and I don't know what is SNMP.

**Comment by Lucienest** [comment]
This Issue should get bumped, as SNMP is really useful.

**Comment by AlmightyJojo** [comment]
I have yet to explore this project but am interested.  However the first thing I did was look if SNMP was included.  Its is very useful.  But does take time to get configured and learn device OIDs.

I currently use Syncro RMM but have been looking for alternative solutions.  The way Syncro does it is their main Windows based monitor agent installed at customer's local site gets designated as a SNMPv2 monitoring agent.  It then does over LANs to query SNMP OIDs then reports back cloud via the designated monitoring agent.

Here's maybe some inspiration to consider adding SNMP.  Lots good monitoring and info.
Pics of Poweredge IDRAC8, N2024 2x switch stack, and APC UPS.

https://i.imgur.com/nSsMz7P.png
https://i.imgur.com/SnBbQGT.png
https://i.imgur.com/2H9l6Kx.png

**Comment by phinnay** [comment]
Bumping this as it'd be very useful for how I am using UptimeKuma. 

@louislam Respectfully disagree that this is outside the scope of the application - SNMP is simply a way to query for data, the logic as to what that data does in UptimeKuma would be where ya'd want to be careful. I agree getting into the weeds with complicated checks is better suited to other products, (Icinga, Nagios, etc...) however querying SNMP for simple values like toner level or if a switch port is up / down is no more out of scope than the existing SQL monitoring that operates the same way. 

I'd propose to literally copy how the SQL monitor works but for SNMP. Simply providing a red/yellow/green status and a graph over time of the queried value would be of tremendous value, even for home users. 

Thanks again for a wonderful product and the time that has gone into it. I wish I was better with code to help implement this feature myself.

**Comment by mattv8** [comment]
I'd also like to bump this. The perfect use-case for me would be to monitor the status of our security cameras- whether they are recording or not. While I can simply set up a ping-test for all of our cameras, this doesn't actually indicate to me that they are successfully recording to the NVR. We use a Lorex NVR system that has an SNMP interface with a query-able status for each camera.

**Comment by Lucienest** [comment]
Bump, please add support for it

**Comment by wglenn01** [comment]
Yes again please.

**Comment by charettemtb** [comment]
Bump +1

**Comment by Lucienest** [comment]
Bump + 2

**Comment by ThaBlazee** [comment]
Bump +3

**Comment by chakflying** [comment]
You can upvote the original post instead of commenting. Comment won't get you there faster.

**Comment by Lucienest** [comment]
Please Bump + 4

**Comment by CommanderStorm** [comment]
@Lucienest @ThaBlazee @wglenn01 @charettemtb

Please refrain from posting `+1` / requests for updates things on issues, as this makes issue-management harder.
Issues are for discussing what needs to be done how by whom.
[We use 👍🏻 on issues to prioritise work](https://github.com/louislam/uptime-kuma/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

Given that this Project is open source, you could also [provide a Pull Request](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md) (with extensive screenshots, as I would not know where to get a device with `SNMP`) 😉 

PS:
@Lucienest given that the comment from Nelson was not enough, I think being more explicit is in order.
If you continue to +1 without any useful contribution, we will get you reported for abuse/spam.
This is the fourth time you have commented in the spirit of `+1´.
Uptime Kuma is maintained voluntarily by people in their spare time, please respect this.

**Comment by Maikelwolsink** [comment]
Bump +5

**Comment by CommanderStorm** [comment]
@Maikelwolsink Have you read the comment above?

**Comment by vladaurosh** [comment]
> 
> Given that this Project is open source, you could also [provide a Pull Request](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md) (with extensive screenshots, as I would not know where to get a device with `SNMP`) 😉
> 
Cheap Mikrotik switch (40$ new) should get you SNMP.
![image](https://github.com/louislam/uptime-kuma/assets/47857324/dfc33e9a-3d77-46fc-bf35-9a5361df7d81)

**Comment by phinnay** [comment]
> Given that this Project is open source, you could also [provide a Pull Request](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md) (with extensive screenshots, as I would not know where to get a device with `SNMP`) 😉

I just noticed this was getting some activity!

SO many devices support SNMP, if you have a network printer, managed switch... anything there's a high chance of it supporting SNMP. 

Without purchasing anything - Both windows and linux support SNMP reporting out the box. You can enable it in your OS and then query well known OIDs to pull values from your computer 

Example:
https://www.sysadmin.md/snmp-most-useful-linux-oids.html

Hope that helps!!

**Comment by CommanderStorm** [comment]
> exactly what should be supported

Given that monitors which come from the community work better (because the dev providing the implementation understands what SNMP is and has a device to test this on), this would be how to get this feature.
=> A reasonable PR implementing this feature is likely to get merged

See this PR for a somewhat similar monitor: https://github.com/louislam/uptime-kuma/pull/3178
Here is our contribution guide: https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md

**Comment by markosole** [comment]
I am happy to see interest in this, well the same thing I am looking for as well. I am using UptimeKuma for 50+ different services (what's supported) and last piece is (to fill the gaps) SNMP for devices and for a simple metrics. 
I will loook into extending it myself and will share some info. Fingers drossed!

**Comment by NEBELWAENDE** [comment]
Maybe this might be a light path or bridge to get over - i have seen, that "uptime kuma", which i like very much, fires up several ping commands (i use Ping Monitors in Uptime Kuma) as one can see on the console/syslog in the docker image i run. I am wondering if someone could figure out if the net-snmp commands snmpwalk and snmpget resp. snmptable command could be helpfull to get certain snmp values from e.g. printers routers and switches. just a thought that it could be a way ...

**Comment by phinnay** [comment]
> Maybe this might be a light path or bridge to get over - i have seen, that "uptime kuma", which i like very much, fires up several ping commands (i use Ping Monitors in Uptime Kuma) as one can see on the console/syslog in the docker image i run. I am wondering if someone could figure out if the net-snmp commands snmpwalk and snmpget resp. snmptable command could be helpfull to get certain snmp values from e.g. printers routers and switches. just a thought that it could be a way ...

that would make a lot of sense! 

To someone who knows coding this PR could be a literal copy/paste from the SQL monitor but switch out the heavy lifting by passing values to net-snmp

Love the product, hope this gets traction!

**Comment by NEBELWAENDE** [comment]
What about forking uptime kuma and Code SNMP Get via OID ourself?

**Comment by NEBELWAENDE** [comment]
const snmp = require('snmp-native');

// Konfigurieren Sie die SNMP-Parameter
const session = new snmp.Session({ host: 'your_target_host', community: 'your_community_string' });

// Die OID, die Sie abfragen möchten
const oid = [1, 3, 6, 1, 2, 1, 2, 2, 1, 16, 2]; // Beispiel-OID für die Anzahl der empfangenen Pakete

// Führen Sie die SNMP-Abfrage durch
session.get({ oid }, (error, varbinds) => {
  if (error) {
    console.error('Fehler bei der SNMP-Abfrage:', error);
  } else {
    varbinds.forEach((varbind) => {
      console.log(`OID: ${varbind.oid.join('.')} = ${varbind.value}`);
    });
  }

  // Beenden Sie die SNMP-Sitzung
  session.close();
});

**Comment by NEBELWAENDE** [comment]
ich würde gerne die Anzahl der Empfangenen Pakete gegen einem Schwellwert überprüfen.

**Comment by NEBELWAENDE** [comment]
In this example, the number of received packets is compared to the threshold value. If the number of received packets exceeds the threshold, it will display the message "Received packets exceeded the threshold of {threshold}" and you can perform your desired action at that point.

Adjust the threshold value as needed to set your specific threshold. When the threshold is exceeded, you can, for example, send an email, generate a notification, or take other actions relevant to your application.

**Comment by NEBELWAENDE** [comment]
const snmp = require('snmp-native');

// Configure SNMP parameters
const session = new snmp.Session({ host: 'your_printer_ip', community: 'your_community_string' });

// Define the OIDs for Toner Levels in Percentage as variables (these are example OIDs)
const cyanTonerOID = '1.3.6.1.4.1.2435.2.4.3.1.15.1'; // Example OID for Cyan Toner Level (%)
const magentaTonerOID = '1.3.6.1.4.1.2435.2.4.3.1.15.2'; // Example OID for Magenta Toner Level (%)
const yellowTonerOID = '1.3.6.1.4.1.2435.2.4.3.1.15.3'; // Example OID for Yellow Toner Level (%)
const blackTonerOID = '1.3.6.1.4.1.2435.2.4.3.1.15.4'; // Example OID for Black Toner Level (%)

// Perform SNMP queries for toner levels in percentage for each color
function queryTonerLevel(oid, colorName) {
  session.get({ oid }, (error, varbinds) => {
    if (error) {
      console.error(`Error in ${colorName} Toner SNMP query:`, error);
    } else {
      varbinds.forEach((varbind) => {
        const tonerLevel = varbind.value;
        console.log(`${colorName} Toner Level (%): ${tonerLevel}`);
      });
    }
  });
}

// Query toner levels for all colors
queryTonerLevel(cyanTonerOID, 'Cyan');
queryTonerLevel(magentaTonerOID, 'Magenta');
queryTonerLevel(yellowTonerOID, 'Yellow');
queryTonerLevel(blackTonerOID, 'Black');

// Close the SNMP session (after all queries are completed)
setTimeout(() => {
  session.close();
}, 1000); // Wait for 1 second before closing the session (adjust as needed)

**Comment by NEBELWAENDE** [comment]
In this program, we define separate OIDs for the percentage toner levels for Cyan, Magenta, Yellow, and Black toners. We then perform SNMP queries for each color, retrieve the toner level in percentage, and print the results. Finally, we close the SNMP session after all queries are completed. Please replace 'your_printer_ip' and 'your_community_string' with the correct values for your printer and SNMP setup.

**Comment by vladaurosh** [comment]
Would this help snmp get implemented in Uptime Kuma? I'd do it myself but I'm not developer.

https://github.com/node-red/node-red-nodes/tree/master/io/snmp

**Comment by owlysk** [comment]
Any news? I really want to have this feature to set up notification for printer toner.

**Comment by apio-sys** [comment]
IMHO this sounds out of scope of the product/goal here. Toner levels can easily trigger an alert from the printer itself. I think - but correct me if I'm wrong - that we want to monitor services running from outside our own network (hence SNMP is really out of scope since LAN oriented protocol) to make sure stuff is running for the rest of the world as it should be. That's just my opinion of course. I think the scope of v2 with numerous additions, especially yearly SLA figures is far more important.

**Comment by owlysk** [comment]
Yeah of course, SNMP protocol can do more, than check level of toner in printer. But my use case is monitoring toner in company. 
SNMP can monitor anything you want.
I like to have one monitoring tool for all my needs and actually I need to monitor SNMP values, so I hope this feature will be added soon.

**Comment by CommanderStorm** [comment]
> Any news

No PR has been raised with support for this.
@NEBELWAENDE has posted some code which would need to be fittend into our `monitoring-type`'s and tested that it works.
An example for a PR implementing a `monitoring-type` is https://github.com/louislam/uptime-kuma/pull/3178

Our contribution guide is available here https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md

**Comment by wglenn01** [comment]
I agree, I would LOVE to have snmp monitoring added. Would be SO helpful as
a WISP.

**Comment by AlmightyJojo** [comment]
I recommend those looking for SNMP checkout Zabbix.  Its powerful monitoring system.
It is distributed monitoring solution where you can setup Zabbix proxy on remote sites LAN(s) and have Zabbix Server at your HQ.  Proxy SNMPv2 monitors will only use community string to monitor SNMP devices on the local LAN (SNMPv2 unencrypted).  The Zabbix proxy to server communication can then use encryption PSK or Cert to update the Zabbix server with everything the proxy is monitoring.  Zabbix server then allows for tons a logic to setup alerts.  For example Dell IDRAC server global heath check = 3 (ok) , alert if not equal to 3.

https://github.com/zabbix/zabbix
https://www.zabbix.com/container_images

**Comment by phinnay** [comment]
> I recommend those looking for SNMP checkout Zabbix. Its powerful monitoring system. It is distributed monitoring solution where you can setup Zabbix proxy on remote sites LAN(s) and have Zabbix Server at your HQ. Proxy SNMPv2 monitors will only use community string to monitor SNMP devices on the local LAN (SNMPv2 unencrypted). The Zabbix proxy to server communication can then use encryption PSK or Cert to update the Zabbix server with everything the proxy is monitoring. Zabbix server then allows for tons a logic to setup alerts. For example Dell IDRAC server global heath check = 3 (ok) , alert if not equal to 3.
> 
> https://github.com/zabbix/zabbix https://www.zabbix.com/container_images

I am quite familiar with Zabbix, and use enterprise class monitoring software in my day job all the time. 

Uptime-kuma is an entirely different use-case. If i wanted to get that crazy I'd script up some middleware to dump SNMP values to SQL and check them with the existing uptime-kuma capabilities. Janitoring enterprise software is not something I want to do in my home / spare time. The simplicity of uptime-kuma is attractive for this reason.

My rationale for promoting this feature is so folks can punch in an OID and get *basic* information back, **the same way all the other checks work.** Nobody is suggesting this software needs to do anything or become anything remotely resembling Nagios/icinga/Zabbix/ELK.

**Comment by AlmightyJojo** [comment]
> 
 Janitoring enterprise software is not something I want to do in my home / spare time.
> >

Ok.  Don't use Zabbix if you don't want to.  It's Custodian - Dick

**Comment by CommanderStorm** [comment]
@AlmightyJojo I have reworded your comment to be more [civil](https://github.com/louislam/uptime-kuma/blob/56aa81e337043d4efa76c46dd7dc8320c9688398/CODE_OF_CONDUCT.md#our-pledge). [Being supportive is in the nature of open source](https://opensource.guide/building-community/#dont-tolerate-bad-actors) and I'd ask you to not call people names just because they disagree with you.

**Comment by CommanderStorm** [comment]
(for those wondering: I merged a few comments to make this thread more readable ^^)

**Comment by AlmightyJojo** [comment]
Its half baked movie joke.  No sense of humor in IT - buttercups all around me.

**Comment by Slamkryp** [comment]
It would be really nice with at least minimal snmp abilitys så you can monitor UPS and Switch temp and such things.Skickat från min iPhone XIII1 feb. 2024 kl. 22:27 skrev owlysk ***@***.***>:﻿
Any news? I really want to have this feature to set up notification for printer toner.

**Comment by CommanderStorm** [comment]
@Slamkryp I think my comment has moved so far to the top that reposting is beneficial

> No PR has been raised with support for this. @ NEBELWAENDE has posted some code which would need to be fittend into our `monitoring-type`'s and tested that it works. An example for a PR implementing a `monitoring-type` is #3178
> 
> Our contribution guide is available here https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md

**Comment by Densen2002** [comment]
Good day to all,

I use Uptime Kuma to monitor "complex" devices and at the same time I use "Mikrotik TheDude!" to monitor my local network network devices that can send data via the SNMP protocol.
The SNMP protocol is widely used in operator network devices and other industrial networks.
Routers and access points such as Ruckus, Mikrotik and others provide a lot of useful information about the current state of the network.

It would be fantastic to combine both monitoring systems in Uptime Kuma. Or configure some interface so that Uptime Kuma can collect data via the SNMP protocol.

**Comment by Densen2002** [comment]
I found a way to make friends - Uptime Kuma with Mirkotik without SNMP protocol support.

The task is to monitor in Uptime Kuma the status of Mikrotik and the map of its TheDude networks 


1. in Uptime Kuma we should create a New Monitor, select the PUSH type, and get a unique push URL, define the waiting interval, for example, 60 seconds, and the “polarity” (Flip the status upside down. If the service is reachable, it is DOWN ).

for example`
http://192.168.80.87:3001/api/push/AdSiDEOE6s?status=up&msg=OK&ping=`



2. In Mikrotik, we should create a tool that will “peck” push URL, informing us in Uptime Kuma about what we have in mind - whether it works or does not work.
`
/tool fetch url=http://192.168.80.87:3001/api/push/AdSiDEOE6s?status=up&msg=OK&ping=" keep-result=no`

3. in Uptime Kuma we see that Mikrotik is alive and what we have in mind - whether it works or not.


You can create as many of these Monitors as you like.


The trick is that using the method described above, you can “upload” various events and service states, for example, SNMP OIDs change or Network Watch Alerts, from the TheDude network map to the Uptime Kuma server. And so on.

**Comment by camdenyoung** [comment]
I would strongly support and encourage integration of SNMP.  Its a fundamental and industry standard of monitoring network equipment.

To give some realworld use, for our temporary deployments we monitor not just that a switch is connected (simple ping response), but with SNMP, is an individual port up or down, has it negocitated at the correct link speed, is it full or half duplex.  Any of these could alert to a potential uptime problem about to occur.

We also deploy many microwave links, so for monitoring purposes, we dont just want to know the link is up/connected, but what is the signal strenth (RSSI), what is the throughput, etc.  This is all simply collected via SNMP.   Just because you can get a ping response from the microwave link at the other end, doesnt mean the link isnt about to fail, or is running at an acceptable speed.

We currently use Nagios for our monitoring, which does the job, but its an awful system to actually setup and maintain, and just isnt user friendly for the rest of our less technical team compared to the fantastic interface you've created with Uptime Kuma.

Don't waste time on intergrating OID library's for individual products, leave that to user to enter the values they need to retrieve.  Most quality equipment has this documented anyway and/or there are hundreds of products already out there that let you scan a device to find the OID you require anyway.

**Comment by Slamkryp** [comment]
SNMP would be really nice as most UPS:s, switches, server support SNMP.
Just for the simple thing to get a notfication when UPS is running on battery would rule!

**Comment by CommanderStorm** [comment]
@Slamkryp You are repeating yourself.
Please avoid spamming without additional information.
Being less suttle: If you continue spamming, we will ban you as by our [code of conduct](https://github.com/louislam/uptime-kuma/blob/master/CODE_OF_CONDUCT.md)

The state of this issue has NOT changed from my response to your last post: https://github.com/louislam/uptime-kuma/issues/1675#issuecomment-1925375148

> No PR has been raised with support for this. @ NEBELWAENDE has posted some code which would need to be fittend into our monitoring-type's and tested that it works. An example for a PR implementing a monitoring-type is https://github.com/louislam/uptime-kuma/pull/3178
>
> Our contribution guide is available here https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md

**Comment by owlysk** [comment]
Hello guys,
I decided to try implement this monitor.

I think it should be like this:
You will set up `IP address` of device, SNMP `version` number, `community` string for SNMPv1 & SNMPv2c and `OID` you want to check. One `OID` per monitor. Then I think you should set `control value` and `condition`. For example, I want to check, if `OID` value is `greater than` 30% (30% is my `control value`).
Conditions will be:
- `>` - check integer
- `>=` - check integer
- `=<` - check integer
- `<=` - check integer
- `==` - check integer & string
- `contains` - check if string is in OID value

Monitoring will check OID value against condition. If `OID` value is `>` than `control value` = `UP`. Otherwise it returns `DOWN`. If device will be offline, returns `DOWN`.

Is that behaviour correct?

---
For now I won't implement SNMPv3.

**Comment by camdenyoung** [comment]
Hi @owlysk, yes that's the fundamental concept, but keep in mind an OID from a network appliance can return a string or integer, depending on what element is being monitored.

Going back to my microwave link example from earlier, if I check the state of an ethernet port, the OID response would be the string of 'up' or 'down', but if I check the RSSI (signal strength) that would return a value anywhere between -128 (dBm) to -20 (dBm).

And thanks for looking into this!  Really appreciate it.  I wouldn't rush to implement SNMP v3.  Yes its the newer (more secure protocol because it has auth) but I honestly would hazard a guess most implementations are still on v2, unless crossing wan connections.

**Comment by chakflying** [comment]
Sounds pretty good, but just in case you didn't know, you should take a look at how the json-query monitor works. 

Ideally in the long term, we would want all value comparisons to work with the jsonata syntax, and reuse the database fields for better compatibility (see #3919). I don't think it's worth implementing a custom value comparison functionality just for this monitor.

**Comment by owlysk** [comment]
@camdenyoung can you please provide output from snmpwalk for signal strength? I want to know what response look like.

@chakflying thank you, I didn't know that. I'll take a look.

**Comment by mattv8** [comment]
Hey @owlysk I've started working on implementing this feature. It's nowhere near complete, but I'm dropping my work thus far into PR #4717. I've laid the framework for your control_value logic. I'll continue to work on this next week (I'm assuming I won't have time over the weekend but ya never know).

**Comment by mattv8** [comment]
@owlysk I'm wondering if you'd be able to take a look. I'm basically stuck with a strange timeout issue. See my PR, specifically, [#issuecomment-2083823715](https://github.com/louislam/uptime-kuma/pull/4717#issuecomment-2083823715)

**Comment by camdenyoung** [comment]
Apologies @owlysk, I've been away the last week.  I'm back in the office now if you still need me to send through an snmpwalk?

**Comment by camdenyoung** [comment]
@owlysk below is the result for RSSI (signal strength) from a Siklu EH-710 microwave link as an example

snmpget -v 2c -c public 10.27.2.82 .1.3.6.1.4.1.31926.2.1.1.19.1
SNMPv2-SMI::enterprises.31926.2.1.1.19.1 = INTEGER: -26

**Comment by mattv8** [comment]
@camdenyoung I'm essentially implementing @owlysk's suggested logic for checking SNMP:
> \> - check integer
> \>= - check integer
> =< - check integer
> <= - check integer
> == - check integer & string
> contains - check if string is in OID value

However, one potential issue I foresee is that often times OID's will report values in binary, or OctetString (more on that [here](https://www.npmjs.com/package/net-snmp#:~:text=name%22%0A%7D%3B-,The%20following%20constants%20are%20defined%20in%20this%20object%3A,EndOfMibView,-snmp.TrapType)). So we need to have logic in place to handle those situations.

In my PR I was originally going to use snmp-native, but I am switching to net-snmp because has SNMP walking. As I work on this feature, I wrote an SNMP walking script, which I used to try to parse my Cisco SG-300 switch. I let the walk run for 30 minutes or so and it reported back 211,468 unique OID's parsable in Cisco's default RO library... So I think implementing an SNMP walk for this feature is impractical. Which is frustrating because unless you already know the OID you're wanting to monitor, there isn't some central directory to easily look up, for example, CPU usage on my Cisco switch...

Posting my SNMP walking script for the record:
<details>
  <summary>Click to view code</summary>
  
  ```javascript
  const fs = require('fs');
const snmp = require('net-snmp');

// Configuration
const monitor = {
    ipAddress: 'changeme',
    port: '161',
    communityString: 'public',
    snmpVersion: '2c',
    rootOid: '1.3.6.1',// This is pretty high level
};

// SNMP options
const options = {
    port: monitor.port,
    retries: 1,
    timeout: 5000,
    version: snmp.Version2c,
};

// Create SNMP session
const session = snmp.createSession(monitor.ipAddress, monitor.communityString, options);

// Create writable stream for CSV
const csvStream = fs.createWriteStream('snmp_data.csv');

// Function to handle SNMP walk feed callback
function feedCb(varbinds) {
    for (const varbind of varbinds) {
        if (!snmp.isVarbindError(varbind) && varbind.value !== null && varbind.value !== undefined) {
            console.log(`OID: ${varbind.oid}, Type: ${findType(varbind.type)}, Value: ${varbind.value}`);
            const csvRow = `${varbind.oid},${findType(varbind.type)},${varbind.value}\n`;
            csvStream.write(csvRow);
        }
    }
}

// Function to handle SNMP walk done callback
function doneCb(error) {
    if (error) {
        console.error(error.toString());
    } else {
        console.log('SNMP walk completed.');
    }
}

// Start SNMP walk from the specified OID
session.walk(monitor.rootOid, 20, feedCb, doneCb);

// Function to find the corresponding type name for a given value
function findType(value) {
    const typeName = Object.entries(snmp.ObjectType).find(([key, val]) => val === value);
    return typeName ? typeName[0] : 'Unknown';
}
  ```
</details>

**Comment by camdenyoung** [comment]
As you've just discovered, I wouldn't worry about implementing any OID scanning functionality.  There are thousands of tools out there that the user can utilise to find the right OID for the sensor / status to be monitored.  Leave it to the user to know what OID to enter, then only functionality required from the uptime kuma perspective is the logic behind the alert.  Is the value true or false, or is the threshold met or exceeded, etc.

On a side note though, it might be worth having a look over [librenms](https://github.com/librenms/librenms) GitHub.  This is a much higher level (and as a result, significantly more complicated) monitoring solution, but it might help explain how they have implemented the scanning functionality as well as the monitoring side.

**Comment by mattv8** [comment]
Thanks for the tip to [librenms](https://github.com/librenms/librenms), I'll take a look. I'll also continue working on the PR tomorrow morning, but I was able to get basic functionality working at least, so I can sleep tonight...

**Comment by mattv8** [comment]
Alright I finished this monitor and have requested review of the PR. For now, I'm not going to implement any sort of OID discovery-- it's just too complicated. I'm going to assume people know the OID they want to monitor already and can paste it in.

If anyone wants to check out this feature, you can clone my PR and test for yourself.

**Comment by owlysk** [comment]
@camdenyoung Thank you for snmpwalk result. @mattv8 was quicker in implementation, than me :)

@mattv8 I'm testing your PR but I always get error -  `Request timed out`. I'm using SNMPv1.

**Comment by mattv8** [comment]
Hey @owlysk I got lucky and had some spare time. 🙂

I'm currently working through the error you're seeing. There is an issue with how my code is parsing strings: `SNMP Error: toString() radix argument must be between 2 and 36`. Working on this now, should have resolution in an hour or so.

**Comment by mattv8** [comment]
I found the issue and will submit my fix before the end of the day.

**Comment by mattv8** [comment]
@owlysk I have committed my code fixing the issue you were encountering.

**Comment by iamk3** [comment]
@mattv8 Where can I pull the container to do some testing myself?

**Comment by bokkoman** [comment]
Whats the status on this after a year? I'm interested in it as well.

**Comment by CommanderStorm** [comment]
It is released in the beta of v2.0
